### PR TITLE
Refresh token request

### PR DESCRIPTION
**Pull request by thsutton**
Implement basic support for refresh_token grant_type.

### CODE REVIEW

**Comment by thsutton** [comment]
I've force pushed a rebase onto @tvh's more recent master.

The refresh endpoint in the demo server now functions correctly (assuming you don't care about expiry) and the tests exercise that functionality.

**Comment by thsutton** [comment]
Woot! Tests pass.

File: oauth2-server-demo/src/Main.hs [hunk 1]
```diff
@@ -39,16 +43,22 @@ oauth2Conf = do
...
-            (requestClientIDReq, requestClientSecretReq) `S.member` s
+        check RequestPassword{..} st =
+            let s = sCreds st
+            in (requestUsername, requestPassword) `S.member` s
+        check RequestClient{..} st =
+            let s = sCreds st
+            in (requestClientIDReq, requestClientSecretReq) `S.member` s
+        check RequestRefresh{..} st =
+            let r = sRefresh st
+            in isJust $ M.lookup requestRefreshToken r
```
Comment (christian-marie) [inline oauth2-server-demo/src/Main.hs]: I'd have written:

```haskell
M.member requestRefreshToken $ sRefresh st
```

File: oauth2-server-demo/src/Main.hs [hunk 2]
```diff
@@ -39,16 +42,19 @@ oauth2Conf = do
...
-        check RequestPassword{..} s =
-            (requestUsername, requestPassword) `S.member` s
-        check RequestClient{..} s =
-            (requestClientIDReq, requestClientSecretReq) `S.member` s
+        check RequestPassword{..} =
+            S.member (requestUsername, requestPassword) . sCreds
+        check RequestClient{..} =
+            S.member (requestClientIDReq, requestClientSecretReq) . sCreds
+        check RequestRefresh{..} =
+            M.member requestRefreshToken . sRefresh
```
Comment (thsutton) [inline oauth2-server-demo/src/Main.hs]: Yeah, that's much nicer.

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs [hunk 3]
```diff
@@ -143,10 +133,8 @@ checkEndpoint
 checkEndpoint = do
     OAuth2 Configuration{..} <- get
     -- Get the token and scope parameters.
-    token <- fmap (Token . T.decodeUtf8) <$> getParam "token" >>=
-        maybe (missingParam "token") return
-    _scope <- fmap T.decodeUtf8 <$> getParam "scope" >>=
-        maybe (missingParam "scope") return
+    token <- Token <$> getRequestParameter "token"
+    scope <- mkScope <$> getRequestParameter "scope"
```
Comment (christian-marie) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: I was going to suggest this but you got there first, nice.

File: oauth2-server/lib/Network/OAuth2/Server.hs [hunk 4]
```diff
@@ -28,16 +28,22 @@ createGrant request = do
...
-             )
-         RequestClient{..} ->
-             ( Just requestClientIDReq
-             , Nothing
-             , fromMaybe (Scope []) requestScope
-             )
+            RequestPassword{..} ->
+                ( requestClientID
+                , Just requestUsername
+                , fromMaybe (Scope []) requestScope
```
Comment (christian-marie) [inline oauth2-server/lib/Network/OAuth2/Server.hs]: Monoid instance seems reasonable.

**Comment by christian-marie** [comment]
Looks good, I'll let you merge when you're ready.

**Pull request merged**